### DISCUSSION TITLE
Libcore .net 8 Docker

### DISCUSSION THREAD

**Discussion by rmpprod** [CryptoPro/libcore#73] [Q&A]
Здравствуйте, у меня есть приложение на .net 8, которое использует LibCore 2024.11.19 и мне нужно залить его в докер.
Отсюда вопрос, есть ли какой-то гайд, как правильно написать докерфайл, чтобы в образе было crypto pro csp, в частности, меня интересует:
1) Как работать с CryptoPro Csp в рамках докер образа
2) Какой тип лицензии нужен, и как собственно лицензировать Csp в докер образе
3) Нужно ли в образе отдельно контейнеры, если да, то как прописывать доступ к ним
4) Требуется ли для работы этой версии LibCore установка CryptoPro .Net, если да, то каким образом он ставиться в докер и как лицензируется
Сам образ пишется под использование на системе Linux, на основе mcr.microsoft.com/dotnet/sdk:8.0 и mcr.microsoft.com/dotnet/aspnet:8.0. Буду очень благодарен, если распишите по подробней :)
Прикладываю свой докерфайл, который написал бегая по форумам, чтобы хоть как-то поднять своё приложение. В нём я уже добавляю нугет пакеты и загружаю CryptoPro Csp в контейнер.
Спасибо за помощь, продублировал обращение на форуме https://www.cryptopro.ru/forum2/default.aspx?g=posts&m=151124#post151124

**Comment by rmpprod** [comment]
![dockerfile](https://github.com/user-attachments/assets/4776a087-8ab8-41c2-b365-e876276d55bf)

**Comment by yodasad** [comment]
Добрый день!

1. Что именно интересует? Использование CSP в docker образах легализовано, начиная с КриптоПро CSP R3 KC1, правила пользования есть в соответствующей документации, входящей в состав дистрибутива: "ЖТЯИ.00101-03 91 09. Руководство администратора безопасности. Виртуальные среды".
2. Если TLS не будет терминироваться в контейнере (а лучше, чтобы этого не было), то подойдёт обычная серверная лицензия на CSP, но требуется учесть, что количество закупленных лицензий должно соответствовать количеству запущенных контейнеров.
3. Наверное, удобнее ключевые-контейнеры хранить отдельно и монтировать при запуске, так можно обеспечить смену ключей без необходимости пересборки образа.
4. LibCore - это и есть КриптоПро .NET, он установится как зависимость самого приложения, лицензируется отдельно от CSP.

Опыт пользователей показал, что использование docker для сколько-нибудь значимых проектов - очень плохая идея, ошибки трудно диагностируемы, как правило, всегда связаны не с работой сервиса, а с инфраструктурой самого docker'а.

**Reply by rmpprod** [reply]
> 4\. LibCore - это и есть КриптоПро .NET, он установится как зависимость самого приложения, лицензируется отдельно от CSP.

Подскажите пожалуйста, каким образом лицензируется именно LibCore на .net 8

**Reply by Fasjeit** [reply]
Лицензия на право использования ПО "КриптоПро .NET" версии 2.0 на одном сервере -- для использования libcore на сервере нужно приобретать именно эту лицензию.
https://cryptopro.ru/order/OrderForm.aspx?catid=aa37ca5a-7cfc-e811-870f-0025900a9ce7
Позиции технической поддержки в этом же разделе.
Для клиентских рабочих мест продукт не требует приобретения лицензии в отличии от версии для Framework.

подробнее - https://github.com/CryptoPro/libcore/issues/22
